data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/actions/CombineAggregateActionTest.java

```diff
@@ -0,0 +1,109 @@
+        firstEventMap.put("array", Arrays.asList(1, 2, 3));
+        eventMaps.add(firstEventMap);
+        events.add(buildEventFromMap(firstEventMap));
+
+        final Map<String, Object> secondEventMap = new HashMap<>();
+        secondEventMap.put("string", "secondEventString");
+        secondEventMap.put("number", 2);
+        eventMaps.add(secondEventMap);
+        events.add(buildEventFromMap(secondEventMap));
+
+        final Map<String, Object> thirdEventMap = new HashMap<>();
+        thirdEventMap.put("string", "thirdEventString");
+        thirdEventMap.put("double", 2.1);
+        eventMaps.add(thirdEventMap);
+        events.add(buildEventFromMap(thirdEventMap));
+
+    }
+
+    private Event buildEventFromMap(final Map<String, Object> eventMap) {
+        return JacksonEvent.builder()
+                .withEventType("event")
+                .withData(eventMap)
+                .build();
+    }
+
+    @Test
+    void handleEvent_with_null_group_state_should_return_correct_AggregateResponse() {
+        final AggregateActionResponse aggregateActionResponse = combineAggregateAction.handleEvent(events.get(0), null);
+
+        assertThat(aggregateActionResponse.getEvent().isPresent(), equalTo(false));
+        assertThat(aggregateActionResponse.isCloseWindowNow(), equalTo(false));
+    }
+
+    @Test
+    void combining_with_single_event_should_result_in_that_same_event_being_returned_on_concludeGroup() {
+        final Map<Object, Object> groupState = new HashMap<>();
+        final AggregateActionResponse aggregateActionResponse = combineAggregateAction.handleEvent(events.get(0), groupState);
+        assertThat(aggregateActionResponse.getEvent().isPresent(), equalTo(false));
+        assertThat(aggregateActionResponse.isCloseWindowNow(), equalTo(false));
+
+        final Optional<Event> result = combineAggregateAction.concludeGroup(groupState);
+        assertThat(result.isPresent(), equalTo(true));
+        assertThat(result.get().toMap(), equalTo(events.get(0).toMap()));
+        assertThat(result.get().getMetadata().getEventType(), equalTo(CombineAggregateAction.COMBINED_EVENT_TYPE));
+    }
+
+    @Test
+    void combining_with_multiple_events_should_combine_into_group_state_correctly_and_return_correct_event_on_concludeGroup() {
+        final Map<Object, Object> groupState = new HashMap<>();
+        final Map<String, Object> expectedResult = new HashMap<>();
+        for (final Map<String, Object> eventMap : eventMaps) {
+            expectedResult.putAll(eventMap);
+        }
+
+        final AggregateActionResponse firstAggregateActionResponse = combineAggregateAction.handleEvent(events.get(0), groupState);
+        assertThat(firstAggregateActionResponse.getEvent().isPresent(), equalTo(false));
+        assertThat(firstAggregateActionResponse.isCloseWindowNow(), equalTo(false));
+
+        final AggregateActionResponse secondAggregateActionResponse = combineAggregateAction.handleEvent(events.get(1), groupState);
```

Comment:
In line with my comment above on testing one function, this can be simplified.

One test will test `handleEvent` adds an event to an empty `groupState`. A second test will start with a `groupState` with some events. Then it tests that `handleEvent` adds new events. These tests combined tell you recursively that further combinations will work.

Reply:
My thought was that instead of making the AggregateAction creator check if groupState is empty and add the Event to it, I would handle this logic in the AggregateProcessor. Otherwise every single AggregateAction will have to copy and paste the following code.
```
if (groupState.size() == 0)
     addEventToGroupState
```

I suppose this is fine to have to do, but is it easier to not have to repeat this every time on the AggregateAction side? I guess it does give the AggregateAction creator the freedom to not add an Event to the groupState (even though this wouldn't really make sense to do)

Reply:
I'm not suggesting changing the code. The current code should work whether it is the first event in a group or not.

But, tests are different. The tests should cover the important different cases. So what I'm suggesting is that we have:
1. A test case to represent the first event in a group. Yes, the `groupState` will be empty in the test setup.
2. A test case to represent adding an event to a group where some other event (or events) have already been added. This test will setup a `groupState` with data.

If both of these tests pass, then we can be reasonably certain that all sorts of combinations (two events, three events, etc.) will work.